weave/flow/prompt/message.py

```diff
@@ -0,0 +1,183 @@
+
+        part = ""
+        escaped = False
+        in_placeholder = False
+        for char in s:
+            if char == "\\":
+                escaped = True
+                continue
+            if escaped:
+                part += char
+                escaped = False
+                continue
+            if char == "{":
+                in_placeholder = True
+                if part:
+                    parts.append(part)
+                    part = ""
+                continue
+            if char == "}":
+                in_placeholder = False
+                parts.append(Placeholder.from_str(part))
+                part = ""
+                continue
+            part += char
+
+        if escaped:
+            raise ValueError("String cannot end with escape character")
+        if in_placeholder:
+            raise ValueError("Unterminated placeholder")
+        if part:
+            parts.append(part)
+        return MessageParts(parts)
+
+    def __repr__(self) -> str:
+        return f"MessageParts(parts={self.parts})"
+
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, MessageParts):
+            return NotImplemented
+        return self.parts == other.parts
+
+    def as_rich_str(self) -> str:
+        rich_parts = []
+        for part in self.parts:
+            if isinstance(part, Placeholder):
+                rich_parts.append(part.as_rich_str())
+            else:
+                rich_parts.append(part)
+        return "".join(rich_parts)
+
+
+# TODO: Restrict more
+
+Detail = Literal["low", "high", "auto"]
+
+
+class Message:
+    content: MessageParts
+    messages: list[dict]
+    role: str
+    templating: Templating
+
+    def __init__(
+        self,
+        content: Optional[Union[str, dict, list]] = None,
+        role: str = "user",
+        templating: Templating = "weave",
+    ):
+        self.templating = templating
+        self.role = role
+        if isinstance(content, list):
+            self.content = MessageParts.from_list(content)
+        elif isinstance(content, dict):
+            # Would it be better to throw if content is missing?
+            self.role = content.get("role", role)
+            self.content = MessageParts([content.get("content", "")])
+        else:
+            self.content = MessageParts.from_str(content or "", templating)
+        # TODO: need to be able to init messages too
+        self.messages = []
+
+    def bind(self, params: Optional[Params] = None) -> dict[str, Any]:
+        params = params or {}
+        bound: dict[str, Any] = {"role": self.role}
+        if self.content:
+            bound["content"] = self.content.bind(params)
+        if self.messages:
+            # TODO: Support messages with placeholders
+            bound["content"] = self.messages
+        return bound
+
+    def add_text(self, text: str) -> "Message":
+        self.messages.append({"type": "text", "text": text})
+        return self
```

Comment:
We should add `add_image` which converts it to a data uri and calls `add_image_url`.